pkg/alertmanager/operator.go

```diff
@@ -131,6 +167,38 @@ func New(ctx context.Context, c prometheusoperator.Config, logger log.Logger, r
 		return nil, errors.Wrap(err, "error creating alertmanager informers")
 	}
 
+	o.alrtCfgInfs, err = informers.NewInformersForResource(
+		informers.NewMonitoringInformerFactories(
+			o.config.Namespaces.AlertmanagerAllowList,
```

Comment:
Prometheus operator would only watch for `AlertmanagerConfig` resources that live in the same namespace(s) than `Alertmanager` resources. If we want to support a different mode, we'd need to introduce an extra parameter/flag (e.g. `--alertmanagerconfig-instance-namespaces=...`). But this isn't going to work if you want to reconcile `Alertmanager` resources from one namespace with `AlertmanagerConfig` from all namespaces.

@s-urbaniak @brancz thoughts?

Reply:
Taking precedence of the thanos ruler sub-operator code: The thanos ruler operator loop watches `PrometheusRule` objects in `o.config.Namespaces.AllowList`, not `o.config.Namespaces.ThanosRulerAllowList`., because `--thanos-ruler-instance-namespaces` is supposed to restrict namespaces that _deploy_ a whole thanos ruler.

I think the same rule applies here. `--alertmanager-instance-namespaces`, which translates to `o.config.Namespaces.AlertmanagerAllowList` should restrict namespaces that _deploy_ a whole alertmanager.

`AlertmanagerConfig` resources on the other hand only control a part of alertmanager configuration, not a whole deployment. This would be in line with the logic elsewhere. Users could restrict watching `AlertmanagerConfig` resources using `--namespaces` and `--deny-namespaces` and have finer grained control of namespaces which deploy an alertmanager instance using `--alertmanager-instance-namespaces`.

In the future we can add finer grained CLI options for concrete resources (not only `AlertmanagerConfig`) as a whole new set of settings.

TLDR: I suggest referencing `o.config.Namespaces.AllowList` here.

also cc'ing @prometheus-operator/prometheus-operator-reviewers for another set of opinions.

Reply:
:+1: to what @s-urbaniak outlined

Reply:
Done, thanks for the guidance! :slightly_smiling_face: